### PR TITLE
WIP for #2987 - First pass implementation for standardised impact functions

### PR DESCRIPTION
**Pull request by ismailsunni**
## IF Implementation
Here is my plan (high level) of implementation.
if class revisited

1. Red box : Pre Processing
2. Black box : IF setup / setup_algorithm (based on hazard and exposure layer)
3. Orange box : Post processing
4. Green box : The algorithms (currently, they are separated class)
5. Blue box : The Impact Function class.

Need to be added:

1. Pre Processing
  1. Clipping
  2. Converting hazard layer to classified vector and add the value map

2. Post Processing
  1. All current post processing (gender, minimum needs, etc)
  2. Need to think where we put the action and notes since we are not using json anymore here. The only logical option is metadata.

Out side the blue box is the presentation layer. We need:

1. Template machine (jinja2)
2. Report format

### Dock interaction:
- Basically, we setup the IF based on layers chosen. Perhaps we can attach bind the IF object to the dock (?)
- IF name or the title (might by affected, flooded, etc) is generated on the fly

### Wizard Interaction
- Need to think about the flow

### Other system
- The other system should only call Impact Function class and give the required parameter

## Home work
- Hazard and exposure metadata which will be the driver of the IF

PS : 
Tim's drawing:

After heavy discussing with @timlinux @Gustry and @lucernae , we get this more details diagram for the new impact function architecture:
inasafe new impact function

You can see the realtimeboard of the diagram in here : https://realtimeboard.com/app/board/o9J_k1JfWJI=/

cc @timlinux @Gustry @lucernae 
please comment :D

### CODE REVIEW

File: safe/new_impact_function/test/test_impact_function.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+from safe.test.utilities import get_qgis_app, standard_data_path
+QGIS_APP, CANVAS, IFACE, PARENT = get_qgis_app()
+
+from safe.new_impact_function.impact_function import ImpactFunction
+
+from qgis.core import QgsVectorLayer
+
+
+def read_json_flow(json_path):
+    """Helper method to read json file that contains a scenario
```
Comment (Gustry) [inline safe/new_impact_function/test/test_impact_function.py]: Missing dot :)

File: safe/new_impact_function/impact_function.py [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+        # Aggregation Preparation
+        if not self.aggregation:
+            impact_function_state['aggregation']['provided'] = False
+            if not self.actual_extent:
+                self._actual_extent = self.exposure.extent()
+            # Create aggregation layer from bbox = QgsVectorLayer()
+            impact_function_state['aggregation']['process'].append(
+                'Convert bbox aggregation to polygon layer with keywords')
+            aggregation_layer = QgsVectorLayer(
+                "Polygon?crs=epsg:4326", "aggregation", "memory")
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Should we use 4326 ? As we are going to reproject to exposure CRS I think.

File: safe/new_impact_function/test/test_impact_function.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+        impact_function = ImpactFunction()
+        impact_function.exposure = exposure_layer
+        impact_function.hazard = hazard_layer
+        self.assertEqual(
+            impact_function.algorithm, impact_function.line_algorithm)
+        self.assertEqual(impact_function.name, 'Flood Polygon on Road Line')
+        self.assertEqual(impact_function.title, 'be affected')
+
+    @unittest.expectedFailure
```
Comment (Gustry) [inline safe/new_impact_function/test/test_impact_function.py]: We should add a reason why it's an expected failure

File: safe/new_impact_function/test/test_impact_function.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            self.assertDictEqual(result, expected)
+
+        path = standard_data_path('scenario')
+        json_files = [os.path.join(path, f) for f in os.listdir(path) if os.path.isfile(os.path.join(path, f))]
+        for json_file in json_files:
+            test_scenario(json_file)
+
+
+
+
```
Comment (Gustry) [inline safe/new_impact_function/test/test_impact_function.py]: Is-it PEP8 ?

File: safe/test/data/scenario/scenario_1.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (Gustry) [inline safe/test/data/scenario/scenario_1.json]: Can we give a more explicit name to the scenario ? haz + exp

Comment (ismailsunni) [inline safe/test/data/scenario/scenario_1.json]: Do you mean change the hazard file name or how?

Comment (Gustry) [inline safe/test/data/scenario/scenario_1.json]: The file name. "scenario_1.json" is not very explicit, if we have many scenarii, then it's gona be difficult to know.

File: safe/new_impact_function/impact_function.py (hunk 2 above)
Comment (ismailsunni) [inline safe/new_impact_function/impact_function.py]: Updated in the newer commit

File: safe/new_impact_function/test/test_impact_function.py (hunk 4 above)
Comment (ismailsunni) [inline safe/new_impact_function/test/test_impact_function.py]: Fixed in the newer commit.

File: safe/test/data/scenario/scenario_1.json (hunk 5 above)
Comment (ismailsunni) [inline safe/test/data/scenario/scenario_1.json]: Changed to polygon_hazard_point_exposure.json

File: safe/new_impact_function/test/test_impact_function.py (hunk 3 above)
Comment (ismailsunni) [inline safe/new_impact_function/test/test_impact_function.py]: Added in the comment

File: safe/new_impact_function/impact_function.py [hunk 6]
```diff
@@ -0,0 +1,829 @@
...
+        formula = formula.replace(key, str(value))
+    return eval(formula)
+
+
+class ImpactFunction(object):
+    """Impact Function."""
+
+    def __init__(self):
+        self._hazard = None
+        self._hazard_keyword = {}
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: We are going to use monkey patching to add  'keywords' to the 'QgsMapLayer'.
In the exposure, hazard and aggregation setters, you should monkey patch keywords.

The workflow will be to read keywords at the beginning of the IF and write them into the layer.keywords as a dictionary. Then in the impact function, we will only work with this monkey patching to insert/update/delete keywords. Only at the end of the impact function, we should use keyword_io to write them on disk.

So we should remove self._hazard_keyword etc

See PR https://github.com/inasafe/inasafe/pull/3219

File: safe/new_impact_function/impact_function.py [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+    @hazard.setter
+    def hazard(self, layer):
+        """Setter for hazard layer property.
+
+        :param layer: Hazard layer to be used for the analysis.
+        :type layer: QgsMapLayer
+        """
+        self._hazard = layer
+        try:
+            self._hazard.keyword = KeywordIO().read_keywords(layer)
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: I started to use `keywords`, not `keyword`. It seems more logical as we have many keyword ;-)

File: safe/new_impact_function/impact_function.py [hunk 8]
```diff
@@ -0,0 +1,766 @@
...
+    @exposure.setter
+    def exposure(self, layer):
+        """Setter for exposure layer property.
+
+        :param layer: exposure layer to be used for the analysis.
+        :type layer: QgsMapLayer
+        """
+        self._exposure = layer
+        try:
+            self._exposure.keyword = KeywordIO().read_keywords(layer)
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Same for `keywords`

File: safe/new_impact_function/impact_function.py [hunk 9]
```diff
@@ -0,0 +1,766 @@
...
+    @aggregation.setter
+    def aggregation(self, layer):
+        """Setter for aggregation layer property.
+
+        :param layer: aggregation layer to be used for the analysis.
+        :type layer: QgsMapLayer
+        """
+        self._aggregation = layer
+        try:
+            self._aggregation.keyword = KeywordIO().read_keywords(layer)
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Same for `keywords`

File: safe/new_impact_function/impact_function.py [hunk 10]
```diff
@@ -0,0 +1,766 @@
...
+                post_processor)
+            self.set_state_process(
+                'post_processor',
+                'Post processor for %s %s.' % (
+                    post_processor['name'], post_processor_output))
+
+    def run(self):
+        self.preprocess()
+        self.run_algorithm()
+        self.post_process()
```
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Somewhere, at the end of the process, we need to write the keywords back using keyword_io.

**Comment by Gustry** [comment]
Good work, do you think you will merge soon ?
We should try to keep our PR atomic and small.

File: safe/new_impact_function/impact_function.py (hunk 7 above)
Comment (ismailsunni) [inline safe/new_impact_function/impact_function.py]: Done.

File: safe/new_impact_function/impact_function.py (hunk 9 above)
Comment (ismailsunni) [inline safe/new_impact_function/impact_function.py]: Done also.

File: safe/new_impact_function/impact_function.py (hunk 10 above)
Comment (ismailsunni) [inline safe/new_impact_function/impact_function.py]: I have added it. But comment it out since it will make the test failed because there is no impact layer created.

**Comment by ismailsunni** [comment]
@Gustry I have addressed your comment. We can merge this if you think it's good to be merged. We can add more code later. I agree with smaller PR.

File: safe/new_impact_function/impact_function.py (hunk 7 above)
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Thanks

File: safe/new_impact_function/impact_function.py (hunk 10 above)
Comment (Gustry) [inline safe/new_impact_function/impact_function.py]: Ok, cool, it's fine like this.

**Comment by Gustry** [comment]
LGTM.
For me you can merge, it's already a big PR, and I may need to work on definitions too.
